rally-domain/src/integration-test/groovy/yakworks/rally/activity/ActivityContactOpTests.groovy

```diff
@@ -192,4 +193,114 @@ class ActivityContactOpTests extends Specification implements DataIntegrationTes
         contacts[3].name == 'Name3'
         contacts[4].name == 'Name4'
     }
+
+    void "op:update and remove with lookup"() {
+        setup:
+        List<Contact> contacts = createSomeContacts()
+        Activity activity = setupData(contacts)
+
+        when:
+        def dta = [
+            id      : activity.id,
+            contacts: [
+                op  : 'update',
+                data: [
+                    [sourceId: "Test3"], //should lookup contact from sourceid
+                    [sourceId: "Test4"],
+                ]
+            ]
+        ]
+
+        def updatedAtt = Activity.update(dta)
+        flush()
+        def updatedContacts = updatedAtt.contacts
+
+        then: "should add 2 new contacts, should lookup contacts by sourceid"
+        updatedContacts.size() == 5
+        updatedContacts[3].id == contacts[3].id
+        updatedContacts[4].id == contacts[4].id
+
+        when: "op:remove with lookup"
+        dta = [
+            id      : activity.id,
+            contacts: [
+                op  : 'remove',
+                data: [
+                    [sourceId: "Test1"],//should lookup contact1 and 2 from sourceid and remove
+                    [sourceId: "Test2"],
+                ]
+            ]
+        ]
+
+        updatedAtt = Activity.update(dta)
+        flush()
+        updatedContacts = updatedAtt.contacts
+
+        then: "should add 2 new contacts, should lookup contacts by sourceid"
+        updatedContacts.size() == 3
+    }
+
+    void "replace with lookup"() {
+        when:
+        def cons = createSomeContacts()
+        def att = setupData(cons)
+
+        def dta = [
+            id: att.id,
+            contacts: [
+                [sourceId: "Test3"], //should look up by sourceid
+                [sourceId: "Test4"],
+            ]
+        ]
+
+        def updatedAtt = Activity.update(dta)
+        flush()
+
+        then:
+        updatedAtt.contacts.size() == 2
+        updatedAtt.contacts[0].id == cons[3].id
+        updatedAtt.contacts[1].id == cons[4].id
+    }
+
+    void "test update with json string"() {
+        setup:
+        List<Contact> contacts = createSomeContacts()
+        Activity activity = setupData(contacts)
+
+        when: "data contains ids"
+        def dta = [
+            id  : activity.id,
+            contacts: "[{\"id\": \"${contacts[3].id}\"}, {\"id\": \"${contacts[4].id}\"}]".toString()
```

Comment:
whats wrong with `'[{"id": "${contacts[3].id}"}....'` , why are you escaping quotes and making it so hard to read?